### PR TITLE
added MLPPRobeDecoder

### LINKED ISSUE #2576: MLP-Probing: implementation and experiments
### Describe the task. Describe the task. It can be a feature, a set of experiments, documentation, etc.

In the next days a PR will be merged:
- https://github.com/ecmwf/WeatherGenerator/issues/2563

that adds support for **Linear Probing** of the frozen latent representation via `LinearProbeDecoder` class.

This is a standard SSL-evaluation protocol: train only a `nn.Linear` (optionally preceded by a LayerNorm) on top of a frozen encoder to check whether the target quantity is linearly decodable from the latent space.

A linear probe is a useful but narrow diagnostic — it only tells us whether the information is linearly accessible, not whether (or how much of) it's present at all.

We want to add an **MLP-based probe** (implementing `MLPProbeDecoder`) as a complementary diagnostic, so we can compare linear vs. MLP probe accuracy. It will be probably a shallow MLP, so it remains a representation diagnostic rather than a second full decoder 

Our idea is the following: a large gap indicates that the representation encodes the target non-linearly, while a small gap indicates the encoder is already producing linearly-readable features (generally the most desirable property)

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

_No response_

### URL to the design document

_No response_

### Area

- [ ] datasets, data readers, data preparation and transfer
- [x] model
- [x] science
- [ ] infrastructure and engineering
- [x] evaluation, export and visualization
- [ ] documentation
- [ ] performance

### PR DESCRIPTION
**Pull request by simone99n**
## Description

MLP-Probing: added MLPProbeDecoder class

## Issue Number

Closes #2576

## Checklist before asking for review

-   [ ] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [ ] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/model/engines.py [hunk 1]
```diff
@@ -1150,3 +1150,51 @@ def forward(self, coords_md, latent_nd, tcs_lens_n1):
...
+        self.with_layer_norm = with_layer_norm
+
+        hidden_dims = hidden_dims if hidden_dims else [latent_dim]
+
+        # Submodule name must not match ``freeze_modules`` regex patterns.
+        self.input_norm = nn.LayerNorm(latent_dim) if with_layer_norm else nn.Identity()
+
+        dims = [latent_dim, *hidden_dims]
+        layers = []
+        for d_in, d_out in zip(dims[:-1], dims[1:], strict=False):
```
Comment (clessig) [inline src/weathergen/model/engines.py:1189]: Is there a reason why we cannot reuse the existing MLP class?

File: src/weathergen/model/engines.py [hunk 2]
```diff
@@ -1150,3 +1150,51 @@ def forward(self, coords_md, latent_nd, tcs_lens_n1):
...
+            layers += [nn.Linear(d_in, d_out), ActivationFactory.get(activation)]
+            if dropout_rate > 0.0:
+                layers.append(nn.Dropout(dropout_rate))
+        layers.append(nn.Linear(dims[-1], out_dim))
+        self.mlp = nn.Sequential(*layers)
+
+    def forward(self, coords_md, latent_nd, tcs_lens_n1):
+        del coords_md
```
Comment (clessig) [inline src/weathergen/model/engines.py:1197]: Let's remove this. It's fine to not use it.

File: src/weathergen/model/engines.py [hunk 3]
```diff
@@ -1150,3 +1150,51 @@ def forward(self, coords_md, latent_nd, tcs_lens_n1):
...
+        latent_md = torch.repeat_interleave(latent_nd, tcs_lens_n1[1:], 0)
```
Comment (clessig) [inline src/weathergen/model/engines.py:1198]: This repeats the 9-cell latent for each (coord-free) decoding?

**Review by clessig — COMMENTED** [review]
Thanks. Some comments inline
